## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#fbe03e75-78c0-4df0-a342-f046141b1089) |
| mint-xl.sh | :x: [more...](#f370a751-0004-4c86-8d5d-26597800b5c4) |
| mint-large-bucket.sh | :x: [more...](#ddd6ad3c-e133-4435-8206-e64f6ee41bb7) |
| mint-compression-dist-xl.sh | :x: [more...](#6d0f7539-5790-4d90-b351-8b87ebfbbe3a) |
<a name="f370a751-0004-4c86-8d5d-26597800b5c4"></a>
### 8419-e15c658/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.52:31200
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fd43f2045d02:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-ajpmoawkmjfvsd",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365832,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15CEF315DB5800AB, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="ddd6ad3c-e133-4435-8206-e64f6ee41bb7"></a>
### 8419-e15c658/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:31164
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 00c7c60945b2:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 55 seconds
(4/14) Running aws-sdk-ruby tests ... done in 18 seconds
(5/14) Running awscli tests ... done in 1 minutes and 21 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 1 minutes and 11 seconds
(8/14) Running minio-dotnet tests ... done in 1 minutes and 35 seconds
(9/14) Running minio-go tests ... FAILED in 3 minutes and 56 seconds
{
  "args": {},
  "duration": 93601,
  "error": "Please reduce your request",
  "function": "testStorageClassMetadataCopyObject()",
  "message": "CopyObject failed on RRS",
  "name": "minio-go: testStorageClassMetadataCopyObject",
  "status": "FAIL"
}

Executed 8 out of 14 tests successfully.
```
<a name="6d0f7539-5790-4d90-b351-8b87ebfbbe3a"></a>
### 8419-e15c658/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:31182
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 26a1bb648737:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 6 seconds
(5/14) Running awscli tests ... done in 2 minutes and 4 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 51 seconds
(8/14) Running minio-dotnet tests ... done in 37 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 14 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 45 seconds
(11/14) Running minio-js tests ... done in 54 seconds
(12/14) Running minio-py tests ... FAILED in 3 minutes and 25 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1396, in test_presigned_get_object_response_headers\n    returned_content_language = response.headers['Content-Language']\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/_collections.py\", line 156, in __getitem__\n    val = self._container[key.lower()]\nKeyError: 'content-language'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2039, in main\n    test_presigned_get_object_response_headers(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1414, in test_presigned_get_object_response_headers\n    raise Exception(err)\nException: 'content-language'\n",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "message": "'content-language'",
  "name": "minio-py:test_presigned_get_object_response_headers",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-b7bf009a-853c-4e77-ab97-b214348f908b",
    "object_name": "6fb64766-58a9-41dc-873f-f0bbff443b2c",
    "presigned_get_object_url": "http://72.28.97.56:31182/minio-py-test-b7bf009a-853c-4e77-ab97-b214348f908b/6fb64766-58a9-41dc-873f-f0bbff443b2c?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20191019%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20191019T045531Z&X-Amz-Expires=5&X-Amz-SignedHeaders=host&response-content-language=en_US&response-content-type=text%2Fplain&X-Amz-Signature=f2a064288c54946ffb6f59b2463cbb25d9100e0035336780f60ab9f974eef79d"
  },
  "duration": 98
}

Executed 11 out of 14 tests successfully.
```
<a name="fbe03e75-78c0-4df0-a342-f046141b1089"></a>
### 8419-e15c658/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.58:32723
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 00b2ca5f1278:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 6 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-jdp9nxyu0mo1vn",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 365699,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 15CEF315DDB9EAA1, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
Error: No such image: minio/minio:8419-e15c658

--- a/cmd/xl-sets.go
+++ b/cmd/xl-sets.go
@@ -1,5 +1,5 @@
 /*
- * MinIO Cloud Storage, (C) 2018 MinIO, Inc.
+ * MinIO Cloud Storage, (C) 2018-2019 MinIO, Inc.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -22,7 +22,6 @@ import (
 	"hash/crc32"
 	"io"
 	"net/http"
-	"sort"
 	"strings"
 	"sync"
 	"time"
@@ -739,78 +738,6 @@ func (s *xlSets) CopyObject(ctx context.Context, srcBucket, srcObject, destBucke
 	}
 	putOpts := ObjectOptions{ServerSideEncryption: dstOpts.ServerSideEncryption, UserDefined: srcInfo.UserDefined}
 	return destSet.putObject(ctx, destBucket, destObject, srcInfo.PutObjReader, putOpts)
-}
-
-// Returns function "listDir" of the type listDirFunc.
-// disks - used for doing disk.ListDir(). Sets passes set of disks.
-func listDirSetsFactory(ctx context.Context, sets ...*xlObjects) ListDirFunc {
-	listDirInternal := func(bucket, prefixDir, prefixEntry string, disks []StorageAPI) (mergedEntries []string) {
-		var diskEntries = make([][]string, len(disks))
-		g := errgroup.WithNErrs(len(disks))
-		for index, disk := range disks {
-			if disk == nil {
-				continue
-			}
-			index := index
-			g.Go(func() error {
-				var err error
-				diskEntries[index], err = disks[index].ListDir(bucket, prefixDir, -1, xlMetaJSONFile)
-				return err
-			}, index)
-		}
-
-		for _, err := range g.Wait() {
-			if err != nil {
-				logger.LogIf(ctx, err)
-			}
-		}
-
-		// Find elements in entries which are not in mergedEntries
-		for _, entries := range diskEntries {
-			var newEntries []string
-
-			for _, entry := range entries {
-				idx := sort.SearchStrings(mergedEntries, entry)
-				// if entry is already present in mergedEntries don't add.
-				if idx < len(mergedEntries) && mergedEntries[idx] == entry {
-					continue
-				}
-				newEntries = append(newEntries, entry)
-			}
-
-			if len(newEntries) > 0 {
-				// Merge the entries and sort it.
-				mergedEntries = append(mergedEntries, newEntries...)
-				sort.Strings(mergedEntries)
-			}
-		}
-
-		return mergedEntries
-	}
-
-	// listDir - lists all the entries at a given prefix and given entry in the prefix.
-	listDir := func(bucket, prefixDir, prefixEntry string) (mergedEntries []string) {
-		for _, set := range sets {
-			var newEntries []string
-			// Find elements in entries which are not in mergedEntries
-			for _, entry := range listDirInternal(bucket, prefixDir, prefixEntry, set.getLoadBalancedDisks()) {
-				idx := sort.SearchStrings(mergedEntries, entry)
-				// if entry is already present in mergedEntries don't add.
-				if idx < len(mergedEntries) && mergedEntries[idx] == entry {
-					continue
-				}
-				newEntries = append(newEntries, entry)
-			}
-
-			if len(newEntries) > 0 {
-				// Merge the entries and sort it.
-				mergedEntries = append(mergedEntries, newEntries...)
-				sort.Strings(mergedEntries)
-			}
-		}
-		return filterMatchingPrefix(mergedEntries, prefixEntry)
-	}
-	return listDir
 }
 
 // FileInfoCh - file info channel
@@ -1711,23 +1638,37 @@ func (s *xlSets) ListBucketsHeal(ctx context.Context) ([]BucketInfo, error) {
 
 // HealObjects - Heal all objects recursively at a specified prefix, any
 // dangling objects deleted as well automatically.
-func (s *xlSets) HealObjects(ctx context.Context, bucket, prefix string, healObjectFn func(string, string) error) (err error) {
-	recursive := true
+func (s *xlSets) HealObjects(ctx context.Context, bucket, prefix string, healObjectFn func(string, string) error) error {
 
-	endWalkCh := make(chan struct{})
-	listDir := listDirSetsFactory(ctx, s.sets...)
-	walkResultCh := startTreeWalk(ctx, bucket, prefix, "", recursive, listDir, endWalkCh)
+	marker := ""
 	for {
-		walkResult, ok := <-walkResultCh
-		if !ok {
+		if globalHTTPServer != nil {
+			// Wait at max 10 minute for an inprogress request before proceeding to heal
+			waitCount := 600
+			// Any requests in progress, delay the heal.
+			for (globalHTTPServer.GetRequestCount() >= int32(globalXLSetCount*globalXLSetDriveCount)) &&
+				waitCount > 0 {
+				waitCount--
+				time.Sleep(1 * time.Second)
+			}
+		}
+
+		res, err := s.ListObjectsHeal(ctx, bucket, prefix, marker, "", 1000)
+		if err != nil {
+			continue
+		}
+
+		for _, obj := range res.Objects {
+			if err = healObjectFn(bucket, obj.Name); err != nil {
+				return toObjectErr(err, bucket, obj.Name)
+			}
+		}
+
+		if !res.IsTruncated {
 			break
 		}
-		if err := healObjectFn(bucket, walkResult.entry); err != nil {
-			return toObjectErr(err, bucket, walkResult.entry)
-		}
-		if walkResult.end {
-			break
-		}
+
+		marker = res.NextMarker
 	}
 
 	return nil

